src/OpenLoco/src/GameCommands/CreateTown.cpp

```diff
@@ -0,0 +1,206 @@
+#include "Date.h"
+#include "Economy/Expenditures.h"
+#include "GameCommands.h"
+#include "GameState.h"
+#include "Localisation/StringIds.h"
+#include "Localisation/StringManager.h"
+#include "Map/SurfaceElement.h"
+#include "Map/TileElement.h"
+#include "Map/TileManager.h"
+#include "Objects/LandObject.h"
+#include "Objects/ObjectManager.h"
+#include "S5/S5.h"
+#include "Ui/WindowManager.h"
+#include "ViewportManager.h"
+#include "World/IndustryManager.h"
+#include "World/StationManager.h"
+#include "World/TownManager.h"
+
+using namespace OpenLoco::Interop;
+using namespace OpenLoco::World;
+
+namespace OpenLoco::GameCommands
+{
+    // 0x00496FE7
+    // TODO: only used here, but move to town manager?
+    static Town* initialiseTown(Pos2 pos)
+    {
+        registers regs;
+        regs.eax = pos.x;
+        regs.ecx = pos.y;
+        call(0x00496FE7, regs);
+
+        if (regs.esi != -1)
+            return reinterpret_cast<Town*>(regs.esi);
+        else
+            return nullptr;
+    }
+
+    // 0x004C5604
+    // TODO: used by other functions; move elsewhere
+    static uint16_t sub_4C5604(Pos2 pos)
+    {
+        registers regs;
+        regs.ax = pos.x;
+        regs.cx = pos.y;
+        call(0x004C5604, regs);
+        return regs.dx;
+    }
+
+    static bool checkSurroundings(Pos2 pos)
+    {
+        auto tile = TileManager::get(pos);
+        auto* surfaceEl = tile.surface();
+        if (surfaceEl->slope() != 0 || surfaceEl->water() != 0)
+        {
+            printf("Slope or water present -- bailing out\n");
+            return false;
+        }
+
+        auto* landObj = ObjectManager::get<LandObject>(surfaceEl->terrain());
+        if (landObj->hasFlags(LandObjectFlags::isDesert | LandObjectFlags::noTrees))
+        {
+            auto dx = sub_4C5604(pos);
+            if (dx < 10 && getGameState().rng.randNext() & 0xFF)
+            {
+                printf("sub_4C5604 does not pass -- bailing out\n");
+                return false;
+            }
+        }
+
+        for (auto& town : TownManager::towns())
+        {
+            auto manhattanDistance = Math::Vector::manhattanDistance(pos, Pos2(town.x, town.y));
+            if (manhattanDistance < 768)
+            {
+                printf("manhattanDistance passes, yay\n");
+                break;
+            }
+        }
+
+        printf("manhattanDistance did not pass, yay\n");
+        return true;
+    }
+
+    // 0x00496C22
+    static uint32_t createTown(const TownPlacementArgs& args, const uint8_t flags)
+    {
+        auto& gameState = getGameState();
+        Pos2 pos = args.pos;
+        Town* newTown = nullptr;
+
+        if (args.pos.x == -1)
+        {
+            bool foundPos = false;
+            for (auto attempts = 200; attempts > 0; attempts--)
+            {
+                uint32_t rand = gameState.rng.randNext();
+                pos = Pos2(((rand & 0xFFFF) * kMapRows) << 5UL, ((rand >> 16) * kMapColumns) << 5UL);
```

Comment:
This line leads to very wrong map positions being generated. Not sure how to improve/fix it yet.